File: backend/app/routes/comissoes.py
"""
Rotas /comissoes — CRUD de comissões de membros da equipe.
"""

from fastapi import APIRouter, HTTPException
from pydantic import BaseModel
from typing import Optional

from app.services.storage_service import comissoes_storage

router = APIRouter()


class ComissaoCreate(BaseModel):
    nome: str
    responsavel: str
    regra: Optional[str] = None       # Descrição da regra / cálculo
    valor: float = 0.0
    competencia: Optional[str] = None
    status: str = "pendente"          # pendente | pago
    observacoes: Optional[str] = None


class ComissaoUpdate(BaseModel):
    nome: Optional[str] = None
    responsavel: Optional[str] = None
    regra: Optional[str] = None
    valor: Optional[float] = None
    competencia: Optional[str] = None
    status: Optional[str] = None
    observacoes: Optional[str] = None


@router.get("")
def list_comissoes():
    comissoes    = comissoes_storage.all()
    total        = sum(float(c.get("valor", 0)) for c in comissoes)
    total_pago   = sum(float(c.get("valor", 0)) for c in comissoes if c.get("status") == "pago")
    total_pend   = sum(float(c.get("valor", 0)) for c in comissoes if c.get("status") != "pago")
    return {
        "comissoes":      comissoes,
        "total":          round(total, 2),
        "total_pago":     round(total_pago, 2),
        "total_pendente": round(total_pend, 2),
        "quantidade":     len(comissoes),
    }


@router.post("")
def create_comissao(body: ComissaoCreate):
    return comissoes_storage.create(body.model_dump())


@router.put("/{id}")
def update_comissao(id: str, body: ComissaoUpdate):
    payload = {k: v for k, v in body.model_dump().items() if v is not None}
    updated = comissoes_storage.update(id, payload)
    if not updated:
        raise HTTPException(status_code=404, detail="Comissão não encontrada")
    return updated


@router.delete("/{id}")
def delete_comissao(id: str):
    if not comissoes_storage.delete(id):
        raise HTTPException(status_code=404, detail="Comissão não encontrada")
    return {"ok": True}

File: functions/src/routes/comissoes.ts
import { Router } from "express";
import { comissoesStorage } from "../lib/firestore";

const router = Router();

function toFloat(v: unknown): number {
  const n = parseFloat(String(v ?? "0"));
  return isFinite(n) ? n : 0;
}

// GET /comissoes
router.get("/", async (_req, res) => {
  try {
    const items = await comissoesStorage.all();
    const total = items.reduce((s, c) => s + toFloat(c.valor), 0);
    const total_pago = items.filter((c) => c.status === "pago").reduce((s, c) => s + toFloat(c.valor), 0);
    const total_pendente = items.filter((c) => c.status !== "pago").reduce((s, c) => s + toFloat(c.valor), 0);
    res.json({ comissoes: items, total: Math.round(total * 100) / 100, total_pago: Math.round(total_pago * 100) / 100, total_pendente: Math.round(total_pendente * 100) / 100, quantidade: items.length });
  } catch (e: unknown) {
    res.status(500).json({ detail: String(e) });
  }
});

// POST /comissoes
router.post("/", async (req, res) => {
  try {
    res.status(201).json(await comissoesStorage.create(req.body));
  } catch (e: unknown) {
    res.status(500).json({ detail: String(e) });
  }
});

// PUT /comissoes/:id
router.put("/:id", async (req, res) => {
  try {
    const updated = await comissoesStorage.update(req.params.id, req.body);
    if (!updated) { res.status(404).json({ detail: "Comissão não encontrada" }); return; }
    res.json(updated);
  } catch (e: unknown) {
    res.status(500).json({ detail: String(e) });
  }
});

// DELETE /comissoes/:id
router.delete("/:id", async (req, res) => {
  try {
    if (!await comissoesStorage.delete(req.params.id)) { res.status(404).json({ detail: "Comissão não encontrada" }); return; }
    res.json({ ok: true });
  } catch (e: unknown) {
    res.status(500).json({ detail: String(e) });
  }
});

export default router;
